app/_src/kubernetes-ingress-controller/upgrade/kic.md

```diff
@@ -4,3 +4,530 @@ type: how-to
+   `CONTROLLER_*` environment variables.
+
+### Changes that do not require user action
+
+1. The `expressions` routing engine is [now the default](#expression-routes-in-db-less-mode) for DB-less
+   environments.
+1. The `console` log format [has changed](#logging-changes).
+
+### Deprecated features that can be changed after upgrading
+
+1. [Convert KongIngress `upstream` fields to KongUpstreamPolicy](#migrating-upstream-configuration).
+1. [Replace `kongCredType` field](#credential-Secrets-now-use-labels) with the
+   `konghq.com/credential` label.
+
+## Change details
+
+### Update CRDs
+
+Helm does not upgrade CRDs automatically. You must apply the 3.x CRDs prior to
+upgrading your releases:
+
+```bash
+kubectl kustomize https://github.com/Kong/kubernetes-ingress-controller//config/crd/?v3.0.0 | kubectl apply -f -
+```
+
+### Migrate KongIngress configuration
+
+3.0 removes support for the KongIngress `route` and `service` fields and their
+subfields.
+
+The KongIngress `upstream` field is now deprecated. The new KongUpstreamPolicy
+resource is now available to hold configuration previously handled by this
+section, and will become the only option for upstream configuration in an
+upcoming release.
+
+#### Migrating route and service settings to annotations
+
+Route (Ingress) and service (Service) configuration fields previously available
+in KongIngress are now all handled via [dedicated annotations][annotations].
+
+For example, if you set the `route.https_redirect_status_code` in a
+KongIngress, you should now use the `konghq.com/https-redirect-status-code`
+annotation on an Ingress or HTTPRoute resource.
+
+##### Migrating header configuration
+
+The [`konghq.com/headers.*` annotation][headers-annotation] uses a special
+format to set headers. The string after the `.` in the annotation name is the
+header name and the annotation value is the header value. For example, to apply
+`x-custom-header-a: example,otherexample` and `x-custom-header-b: example`
+headers to requests, KongIngress configuration would look like:
+
+```yaml
+route:
+  headers:
+    x-custom-header-a:
+    - example
+    - otherexample
+    x-custom-header-b:
+    - example
+```
+
+The equivalent annotation configuration looks like:
+
+```text
+konghq.com/headers.x-custom-header-a: example,otherexample
+konghq.com/headers.x-custom-header-b: example
+```
+
+Note that you cannot apply multiple instances of the same header annotation to
+set multiple header values: you must set the CSV format within a single header.
+
+#### Migrating upstream configuration
+
+The `upstream` section of KongIngress resources contains a complex object that
+does not easily fit in annotations. 3.x uses the new KongUpstreamPolicy
+resource to configure upstream settings. The `spec` field of KongUpstreamPolicy
+is the same as the `upstream` section of KongIngress. For example, if you
+previously used a KongIngress like:
```

Comment:
```suggestion
resource to configure upstream settings. The `spec` field of KongUpstreamPolicy
is similar to the `upstream` section of KongIngress (for exact schema please refer to [KongUpstreamPolicy reference](https://docs.konghq.com/kubernetes-ingress-controller/2.12.x/references/custom-resource-definitions/#kongupstreampolicy). 
For example, if you previously used a KongIngress like:
```

Reply:
Do we know what fields are actually different to call them out here? Trying to visually diff the schemas is probably annoying for as large as that config is.

Reply:
Added more details regarding differences.